packages/react-components/react-dialog/src/components/DialogSurface/useDialogSurface.ts

```diff
@@ -16,39 +27,81 @@ import { isEscapeKeyDismiss } from '../../utils';
  */
 export const useDialogSurface_unstable = (
   props: DialogSurfaceProps,
-  ref: React.Ref<HTMLElement>,
+  ref: React.Ref<DialogSurfaceElement>,
 ): DialogSurfaceState => {
+  const { backdrop, as } = props;
   const modalType = useDialogContext_unstable(ctx => ctx.modalType);
-  const { as = 'div' } = props;
 
-  const contentRef = useDialogContext_unstable(ctx => ctx.contentRef);
+  const dialogRef = useDialogContext_unstable(ctx => ctx.dialogRef);
+  const open = useDialogContext_unstable(ctx => ctx.open);
+  const requestOpenChange = useDialogContext_unstable(ctx => ctx.requestOpenChange);
   const dialogTitleID = useDialogContext_unstable(ctx => ctx.dialogTitleID);
   const dialogBodyID = useDialogContext_unstable(ctx => ctx.dialogBodyID);
-  const requestOpenChange = useDialogContext_unstable(ctx => ctx.requestOpenChange);
 
-  const { modalAttributes } = useModalAttributes({ trapFocus: modalType !== 'non-modal' });
+  const handledBackdropClick = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
+    if (isSlot(props.backdrop)) {
+      props.backdrop.onClick?.(event);
+    }
+    if (modalType === 'modal' && !event.isDefaultPrevented()) {
+      requestOpenChange({
+        event,
+        open: false,
+        type: 'backdropClick',
+      });
+    }
+  });
 
-  const handleRootKeyDown = useEventCallback((event: React.KeyboardEvent<HTMLDivElement>) => {
+  const handleKeyDown = useEventCallback((event: React.KeyboardEvent<DialogSurfaceElementIntersection>) => {
     props.onKeyDown?.(event);
-    if (isEscapeKeyDismiss(event, modalType)) {
-      requestOpenChange({ event, open: false, type: 'escapeKeyDown' });
+
+    if (
+      event.key === Escape &&
+      // `non-modal` should always have Escape key handling
+      // `modal` should only be handled in the case of non native dialog
+      (modalType === 'non-modal' || modalType === 'modal') &&
+      !event.isDefaultPrevented()
+    ) {
+      requestOpenChange({
+        event,
+        open: false,
+        type: 'escapeKeyDown',
+      });
       event.preventDefault();
     }
   });
 
   return {
     components: {
+      backdrop: 'div',
       root: 'div',
     },
-    root: getNativeElementProps(as, {
-      ref: useMergedRefs(ref, contentRef),
-      'aria-modal': modalType !== 'non-modal',
-      'aria-describedby': dialogBodyID,
-      'aria-labelledby': props['aria-label'] ? undefined : dialogTitleID,
-      role: modalType === 'alert' ? 'alertdialog' : 'dialog',
-      ...props,
-      ...modalAttributes,
-      onKeyDown: handleRootKeyDown,
+    backdrop: resolveShorthand(backdrop, {
+      required: open && modalType !== 'non-modal',
+      defaultProps: {
+        'aria-hidden': 'true',
+        onClick: handledBackdropClick,
+      },
     }),
+    root: getNativeElementProps(
+      as ?? 'div',
+      resolveShorthand(props as DialogSurfaceRootSlotProps, {
```

Comment:
this is the first time I've seen `resolveShorthand` used for the root slot, why do you do that here ?

Reply:
Good question... I got not reason, it's definitely pointless.